upgrade crates-index-diff to v13, with 'normalization' support

Byron opened the pull request:
See this PR for details: https://github.com/rust-lang/crates.io/pull/5066.

Note that it brings `gitoxide` in as first step, with `git2` being scheduled for removal towards the end of the year at the latest. Once that happens I am happy to submit another PR and see to removing `git2` here as well.

syphar commented on Cargo.lock:
```diff
@@ -730,13 +804,16 @@ dependencies = [
 [[package]]
 name = "crates-index-diff"
-version = "10.0.0"
+version = "11.0.0"
```
this will probably change to `11.1.1`, right? :)

syphar replied:
(I saw the bug-reports & new releases )

Byron replied:
Indeed, there were a few improvements and the version number was updated in e5320eb8.

syphar commented on Cargo.lock:
```diff
@@ -2121,15 +2686,14 @@ checksum = "e4a24736216ec316047a1fc4252e27dabb04218aa4a3f37c6e7ddbf1f9782b54"
 [[package]]
 name = "nix"
-version = "0.20.2"
+version = "0.20.0"
```
the audit failed (among others) because of the downgrade of this version, because of [RUSTSEC-2021-0119](https://rustsec.org/advisories/RUSTSEC-2021-0119.html). 

Was there a specific reason to downgrade it? 

( I didn't dig into where we actually use the package, we could do that if you need this downgrad)

Nemo157 replied:
I was just trying to find out what was going on here too 😁, the downgrade is caused by `nix 0.20.2` having a limited version range of `bitflags` that isn't compatible with what `git-config` needs:

```console
> cargo update -p nix --precise 0.20.2
error: failed to select a version for `bitflags`.
    ... required by package `nix v0.20.2`
    ... which satisfies dependency `nix = "^0.20.0"` of package `rustwide v0.15.0`
    ... which satisfies dependency `rustwide = "^0.15.0"` (locked to 0.15.0) of package `docs-rs v0.6.0 (/home/nemo157/sources/docs.rs)`
versions that meet the requirements `>=1.1.0, <1.3.0` are: 1.2.1, 1.2.0, 1.1.0

all possible versions conflict with previously selected packages.

  previously selected package `bitflags v1.3.2`
    ... which satisfies dependency `bitflags = "^1.3.2"` of package `git-config v0.7.1`
    ... which satisfies dependency `git-config = "^0.7.1"` (locked to 0.7.1) of package `git-repository v0.23.0`
    ... which satisfies dependency `git-repository = "^0.23.0"` (locked to 0.23.0) of package `crates-index-diff v11.1.1`
    ... which satisfies dependency `crates-index-diff = "^11.0.0"` (locked to 11.1.1) of package `docs-rs v0.6.0 (/home/nemo157/sources/docs.rs)`
```

We likely need to upgrade the `nix` dependency in `rustwide`, the newer versions of `nix` allow using newer versions of `bitflags`.

Byron replied:
Thanks so much for researching this! I wasn't even aware of the possibility of updating individual packages with `cargo update` - nice!

What I find interesting is that `git-config` forces as more recent version even though it most definitely doesn't use any special features of the more recent release. If it helps, I can also downgrade `bitflags` there to 1.2.1, even though waiting for the `rustwide` PR to clear seems like the way to go.

Nemo157 replied:
I forgot to mention here, rustwide 0.15.1 was released with the updated `nix`. If you `cargo update -p rustwide -p nix -p bitflags` it should update just those crates and hopefully get the build passing.

Byron replied:
Thanks a lot for the heads-up! I upgraded both `crates-index-diff` and `rustwide` to their latest version, let's hope for the best.

syphar reviewed (changes requested):
code looks good to me, two comments regarding the dependencies. 

@Nemo157 would you have time for a short manual test? 

( my env is still broken in that regard )

syphar commented on src/index/mod.rs:
```diff
@@ -42,14 +42,15 @@ impl Index {
         let diff = crates_index_diff::Index::from_path_or_cloned_with_options(
             &path,
-            crates_index_diff::CloneOptions {
+            crates_index_diff::index::CloneOptions {
```
This seems to fail in our build-tests: 

```
thread 'main' panicked at 'failed to initialize the index: initialising registry index repository  Caused by:     error reading from the zlib stream; class=Zlib (5)
```

Byron replied:
I can actually reproduce this issue with one of my tools, `criner`, but thus far managed to talk myself into "it's probably something with my local index and nuking it should fix it". Clearly, that is not the case and I am investigating.

It's very strange to see this error as I believe the code for fetching things or cloning didn't change at all, and how can it be that it's entirely within libgit2? Anyway, I shall _hopefully_ find out soon enough :D.

Byron replied:
I could definitely validate that v10 of `crates-index-diff` works, whereas v11 will start streaming the pack but abort with the above message after a couple of seconds. I have a feeling it's less related to zlib, but more to the connection aborting for some reason. `open-ssl` versions are all the same, I tried the vendored openssl feature, as well as `zlib-ng-compat`, but nothing changed.
Thus, the `diff` should reveal the issue, or a binary search.

Byron replied:
It turned out to be a misconfigured zlib which caused the issue. Overall, that's more than 2h of my life I am not going to get back, but at least it's a 2 for 1 deal as I believe `cargo` would have exhibited the same issue 🎉.
And, it's an additional motivator to get `gitoxide` into shape for fetching repositories and finish the transition of `crates-index-diff` 😁.

syphar commented:
@Byron could you ping me here, or re-request the review when this is ready for another round? 

( I would still love to have the manual test by @Nemo157 , or I have to finally take the time to figure out my local setup ;) )

Byron commented:
@syphar I think it's ready. All non-`nix` related tests seem to be passing as well. All I had to change is the version number, all fixes were applied in `crates-index-diff` as well as `git-repository`.

Nemo157 commented:
I updated to v11.1.3 locally and tried running the daemon with registry watcher, that seems to be erroring out when calling `peek_changes`:

```
docsrs-web-1  | 2022/09/01 21:50:37 [ERROR] docs_rs::utils: Failed to get new crates
docsrs-web-1  |
docsrs-web-1  | Caused by:
docsrs-web-1  |     invalid number at line 1 column 2
docsrs-web-1  |
docsrs-web-1  | Stack backtrace:
docsrs-web-1  |    0: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
docsrs-web-1  |              at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/result.rs:2125:27
docsrs-web-1  |       docs_rs::build_queue::BuildQueue::get_new_crates
docsrs-web-1  |              at /build/src/build_queue.rs:199:34
docsrs-web-1  |    1: docs_rs::utils::daemon::start_registry_watcher::{{closure}}
docsrs-web-1  |              at /build/src/utils/daemon.rs:31:27
```

(it would be nice if the errors attached additional context, rather than all being `#[error(transparent)]`, if this is a json error then it could have the path inside the index that failed).

Nemo157 commented:
I thought I had reset `crates-index-diff_last-seen` to the current `HEAD` before running that, but I missed that the docker setup uses an anonymous mount for the index. After execing in and resetting that it seems to be working ok now. The ref it was attempting to diff from was:

```
  crates-index-diff_last-seen        fb8f72861d6 Updating crate `scorpion#0.1.3`
```

Byron commented:
Thanks for sharing! I have improved error messages in [the new release](https://github.com/Byron/crates-index-diff-rs/releases/tag/v11.1.4) which will include the line it failed to decode a crate-version from. That should help.

No matter what the trees are that it diffs, I would expect all of these to yield valid lines of crate-versions, so it's surprising there are lines it cannot interpret at all. If this is something you think can be addressed, then the URL to the crates-index used and the hashes of both commits would be necessary to reproduce it. Thanks.

Nemo157 commented:
I was able to reproduce it with:

```rust
n main() -> anyhow::Result<()> {
    let index = crates_index_diff::Index::from_path_or_cloned_with_options(
        "index",
        crates_index_diff::index::CloneOptions {
            repository_url: "https://github.com/rust-lang/crates.io-index-archive".into(),
            fetch_options: None,
        },
    )?;
    let start = git_hash::ObjectId::from_hex(b"fb8f72861d63ecfd7564ca7eb82f720debe6db4d")?;
    let end = git_hash::ObjectId::from_hex(b"31a1d8c9b1f6851c9b248813b5bb883ba5297883")?;
    let _ = index.changes_between_commits(start, end)?;
    Ok(())
}
```

```
> cargo run --release
[....] long time to clone the archive repo
Error: Failed to decode "---" in update-dl-url.yml as crate version

Caused by:
    invalid number at line 1 column 2
```

(I locally added the filename into the error message too to see where it was failing)

Looks like the issue is this file was added outside the repository schema: <https://github.com/rust-lang/crates.io-index/blob/master/.github/workflows/update-dl-url.yml>

That shouldn't block us using it here, we are past the commit adding that and modifications like that are rare enough that we can just update once it's fixed.

Byron commented:
Thanks so much for your effort!

I have adjusted the fixture generator to [also include the .github folder](https://github.com/Byron/crates-index-diff-rs/blob/8af61f2a20eee72b0e53ae3b6ce22a3a9d52546c/tests/fixtures/make-index-parts.sh#L13) which allows to reproduce the exact same issue, [and fix it](https://github.com/Byron/crates-index-diff-rs/blob/8af61f2a20eee72b0e53ae3b6ce22a3a9d52546c/src/index/diff/delegate.rs#L41-L43). Additionally, and finally 😅, I have added [the filename to json-deserialization errors](https://github.com/Byron/crates-index-diff-rs/blob/8af61f2a20eee72b0e53ae3b6ce22a3a9d52546c/src/index/diff/delegate.rs#L142) which should help debugging without you having to make locally changes, should there be a next time for that.

The fix was published in [the new v11.1.5](https://github.com/Byron/crates-index-diff-rs/releases/tag/v11.1.5) release.

Byron commented:
With CI being green, I think this is ready for another round of reviews. I opted to switch the `Oid` types used in the docs.rs API to their corresponding `gitoxide` versions in anticipation of `gitoxide` being able to fetch the crates index in a month or so, fully replacing `git2` in the process.

syphar commented on Cargo.toml:
```diff
@@ -28,7 +28,7 @@ regex = "1"
 crates-index = { version = "0.18.5", optional = true }
 rayon = { version = "1", optional = true }
-crates-index-diff = "10.0.0"
+crates-index-diff = "12.0.0"
```
this should be `12.1.0`, right?

Byron replied:
Good catch! Fixed in c2cd2b7b70ab0e72a01c90fe7cec01e5d9ed53da.

syphar commented on src/build_queue.rs:
```diff
@@ -49,15 +48,17 @@ impl BuildQueue {
     }
 
-    fn set_last_seen_reference(&self, oid: Oid) -> Result<()> {
+    fn set_last_seen_reference(&self, oid: &crates_index_diff::git::oid) -> Result<()> {
```
nit, consistency: why is this `git::oid` while we return `git::ObjectId` above in the getter?

Byron replied:
The `git::oid` is akin to `std::path::Path` and `git::ObjectId` is akin to `std::path::PathBuf`. Thus it's considered 'better' to take references to object ids if they suffice. Note that `git::ObjectId` is also `Copy` as it's currently only 20 bytes, and I changed the parameter to `git::ObjectId` to resolve the issue in 3cc50c3fc3658fca3726e6149c2ecac4b101ab7a .

syphar reviewed (commented):
thanks for continuing the work here! sorry for the delay reviewing. 

I only have two small comments left.

syphar commented:
cc @Nemo157 do you have any more comments / anything to add?

syphar commented:
@Byron I'm seeing the merge is blocked because of conflicts, could you rebase so I can merge after @Nemo157 gives his go?

Byron commented:
<img width="971" alt="Screen Shot 2022-10-09 at 19 02 40" src="https://user-images.githubusercontent.com/63622/194753225-3158ef3d-67b8-40c1-a8f3-dbf97bdb4264.png">

Great to hear it's ready!

It looks like you want to rebase even though rebasing and squashing and merging seems to work through the github UI. Merging into master seems to work without conflicts, but indeed, when rebasing there are a lot of conflicts in Cargo.toml.

While trying it a few times, also with squashing, I grew a little frustrated so I am here asking if it it's OK to merge into master instead. Looking at the history, I merged master into this branch to resolve conflicts some time ago so any nice history that could have been is no more unless I redo the work. The latter seems easier than rebasing, so if that's desired I'd close this PR and create a new one with a single commit.

I could probably get this done within the next hours if I hear from you with direction.
Thank you

syphar commented:
Thanks for trying, I think if the alternative is a new PR with single commit I can also just to a squash-merge with github, so we can leave it like it is. 

Merge-commits are disabled on the repo.

jyn514 commented:
Squash and merge is fine by me, but @Byron just as a tool you might find useful: when Cargo.lock has merge conflicts I often use `git checkout HEAD Cargo.lock` and then the cargo command I originally ran; that usually fixes the issues without needing manual edits.

Byron commented:
> Squash and merge is fine by me, but @Byron just as a tool you might find useful: when Cargo.lock has merge conflicts I often use `git checkout HEAD Cargo.lock` and then the cargo command I originally ran; that usually fixes the issues without needing manual edits.

Thanks a lot for sharing, I will be sure to remind myself of this advice from now on! Interestingly I had it in the back of my head when sloppily resolving merge conflicts on `Cargo.lock` knowing it will be overwritten anyway, but somehow didn't manage to get to arrive at simply checking it out.

If there is time for a couple of days more, I will upgrade the version of `crates-index-diff` to also clone using `gitoxide` - these clones are faster - in case it's of interest.

Nemo157 commented:
Looks good to me. I'm happy to wait a couple days then I can do some manual testing (currently we never clone the repo as everything runs on the same instance, but we may be migrating some infrastructure around that will cause some new clones so having it faster would be useful).

Byron commented:
That's great to hear! Just now I managed to fully replace `git2` in `crates-index-diff` and took the liberty to do the same here, removing `git2` entirely which hopefully helps compile times to stay more or less the same.

Commit 3423ea5 also contains a note that the reviewer might find interesting as it makes an assumption.

Byron commented on src/index/mod.rs:
```diff
@@ -26,26 +29,29 @@ struct IndexConfig {
+        .tree()?
+        .lookup_entry_by_path("config.json")?
+        .with_context(|| anyhow::anyhow!("registry index missing config"))?
+        .oid
```
In the next release of `git-repository`, these three lines (`oid.attach(&repo).object()`) become just `object()`.

syphar commented:
> Looks good to me. I'm happy to wait a couple days then I can do some manual testing (currently we never clone the repo as everything runs on the same instance, but we may be migrating some infrastructure around that will cause some new clones so having it faster would be useful).

@Nemo157 manual testing would be awesome, so we're safe with a change like this. 
Regarding the new infrastructure work: I remember I tested the separate registry-watcher (`start-registry-watcher`) and saw that it currently also doesn't support cloning the initial repo. I assume that's why we're cloning the index repo in `entrypoint.sh`. 

> That's great to hear! Just now I managed to fully replace git2 in crates-index-diff and took the liberty to do the same here, removing git2 entirely which hopefully helps compile times to stay more or less the same.

Awesome!

Byron commented:
Something I thought would be worth mentioning is that `gitoxide`, similar to `git2`, leaves a small pack after each fetch operation. It's planned to implement `fetch.unpackLimit` to auto-unpack small packs, and to eventually support limited repacking as well (i.e. there are 10k loose objects, let's put those into a pack). I will keep `crates-index-diff` uptodate with relevant releases of `gitoxide` and probably post PRs with updates here if that's OK with you.

The above is probably a long-winded way of saying: nothing changes for you, as I saw that `git gc` is invoked in code from time to time already to alleviate the 'too many small packs issue', but I think that one day `gitoxide` can do the right thing based on existing git-configuration variables to keep repos well maintained automatically.

syphar commented:
@Nemo157 it would be awesome if you could help with a short review & a manual test here.

syphar commented:
ping @Nemo157 for a test, only for the case that this got lost. I totally understand if there wasn't time.

Nemo157 commented:
I've got this running on my server now. Initially seems to be working fine, it added a bunch of crates from the ~100 commit old last-seen-reference I set and has added a few more since. I'll leave it with the registry watcher enabled for now to see if anything happens.

Nemo157 commented:
I haven't seen anything weird skimming the last 13 hours of logs, so this seems to be working fine.

syphar commented:
@Byron could you merge master & fix the conflicts? Then I'll squash-merge this PR

Byron commented:
It's finally done 🎉.
